misc/formatter/modules/formatter-core/src/main/java/org/ballerinalang/formatter/core/NodeFormattingMetrics.java

```diff
@@ -0,0 +1,116 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.ballerinalang.formatter.core;
+
+import io.ballerinalang.compiler.syntax.tree.BuiltinSimpleNameReferenceNode;
+import io.ballerinalang.compiler.syntax.tree.Node;
+import io.ballerinalang.compiler.syntax.tree.NonTerminalNode;
+import io.ballerinalang.compiler.syntax.tree.ObjectFieldNode;
+import io.ballerinalang.compiler.syntax.tree.SyntaxKind;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+
+/**
+ * Class that stores the formatting metrics for a particular node.
+ */
+class NodeFormattingMetrics {
+
+    private Node node;
+    private NonTerminalNode parent;
+    private NonTerminalNode grandParent;
+
+    NodeFormattingMetrics(Node node) {
+
+        this.node = node;
+        this.parent = node.parent();
+        if (parent != null) {
+            this.grandParent = parent.parent();
+        } else {
+            this.grandParent = null;
+        }
+        setNodeMetrics();
+    }
+
+    private void setNodeMetrics() {
+        if (node instanceof BuiltinSimpleNameReferenceNode) {
+            handleBuiltinSimpleNameReferenceNode();
+        }
+    }
+
+    boolean handleBuiltinSimpleNameReferenceNode() {
+        boolean addSpaces = true;
+        ArrayList<SyntaxKind> parentNodes = new ArrayList<>(
+                Arrays.asList(
+                        SyntaxKind.FUNCTION_CALL,
+                        SyntaxKind.TYPE_TEST_EXPRESSION,
+                        SyntaxKind.TYPE_PARAMETER,
+                        SyntaxKind.TYPE_CAST_PARAM,
+                        SyntaxKind.UNION_TYPE_DESC,
+                        SyntaxKind.XML_TYPE_DESC));
+        if (parent != null && (parentNodes.contains(parent.kind()) || grandParent != null &&
+                ((parent.kind().equals(SyntaxKind.TYPED_BINDING_PATTERN) &&
+                        grandParent.kind().equals(SyntaxKind.FOREACH_STATEMENT)) ||
+                (parent.kind().equals(SyntaxKind.UNION_TYPE_DESC) &&
+                        grandParent.kind().equals(SyntaxKind.PARAMETERIZED_TYPE_DESC)) ||
+                (parent.kind().equals(SyntaxKind.OBJECT_FIELD) &&
+                        ((ObjectFieldNode) parent).visibilityQualifier().isPresent()) ||
+                (grandParent.kind().equals(SyntaxKind.LOCAL_VAR_DECL) &&
+                        grandParent.children().get(1).equals(parent)) ||
+                grandParent.kind().equals(SyntaxKind.FROM_CLAUSE) ||
+                grandParent.kind().equals(SyntaxKind.PARAMETERIZED_TYPE_DESC)))) {
```

Comment:
Condition of this is-block is a tad too big. Let's extract it to a separate function.

We can simply do: `addSpaces = newlyAddedFunction();`

Reply:
Thanks for the input, planning on refactoring and simplifying the logic separately in a different PR.